### PR TITLE
Modify interaction model for device and rule flyouts

### PR DESCRIPTION
**Pull request by mechaffin**
# Type of change? 

- [ ] Bug fix
- [ ] New feature
- [x] Enhancement
- [ ] Breaking change (breaks backward compatibility)

# Description, Context, Motivation 

1. Flyouts should stay open once their action has been completed (Save, Apply, Delete, etc).
2. The redux store should be updated directly instead of causing a refresh from the server.
3. The SummarySection should show a round indicator when the action is in progress. Then show a checkmark when completed successfully. If an error occurs instead, show the error message.
4. If the user makes changes to fields on the flyout, then the action button should be re-enabled.

This applies to all device and rule flyouts.

### CODE REVIEW

File: public/locales/en/translations.json [hunk 1]
```diff
@@ -220,6 +220,8 @@
         "title": "Delete",
         "header": "Delete device",
         "description": "This action will delete the devices permanently from the platform.",
+        "confirmNo" : "No, do not delete the selected physical devices.",
```
Comment (spryor) [inline public/locales/en/translations.json]: nit: space before the `:` here and the next line

File: src/components/pages/devices/flyouts/deviceDelete/deviceDelete.js [hunk 2]
```diff
@@ -46,8 +49,11 @@ export class DeviceDelete extends Component {
   }
 
   componentWillReceiveProps(nextProps) {
-    if (nextProps.devices && (this.props.devices || []).length !== nextProps.devices.length) {
-      this.populateDevicesState(nextProps.devices);
+    const nextDevices = this.ensureDevicesExist(nextProps.devices);
+    const curDevices = this.state.devices || [];
```
Comment (spryor) [inline src/components/pages/devices/flyouts/deviceDelete/deviceDelete.js]: Isn't `state.devices` guaranteed to always be initialized to an array (as on line 34)?

File: src/components/pages/devices/flyouts/deviceDelete/deviceDelete.js [hunk 3]
```diff
@@ -61,13 +67,49 @@ export class DeviceDelete extends Component {
     if (this.subscription) this.subscription.unsubscribe();
   }
 
+  /*
+  Because the ag-grid doesn't update its state for selected items when something is removed from its props/rowData,
```
Comment (spryor) [inline src/components/pages/devices/flyouts/deviceDelete/deviceDelete.js]: Check that this is still true with the changes to the data model that went in yesterday.

File: src/components/pages/devices/flyouts/deviceDelete/deviceDelete.js [hunk 4]
```diff
@@ -61,13 +67,49 @@ export class DeviceDelete extends Component {
...
+  we need to take the props.devices we get here and ensure it still exists in the redux store.
+  */
+  ensureDevicesExist = (devices = []) => {
+    return devices.reduce(
```
Comment (spryor) [inline src/components/pages/devices/flyouts/deviceDelete/deviceDelete.js]: Use the implicit return

File: src/components/pages/devices/flyouts/deviceDelete/deviceDelete.js [hunk 5]
```diff
@@ -61,13 +67,49 @@ export class DeviceDelete extends Component {
...
+      (existingDevices, entity) => {
+        if (this.props.deviceEntities[entity.id]) {
```
Comment (spryor) [inline src/components/pages/devices/flyouts/deviceDelete/deviceDelete.js]: `if (entity.id in this.props.deviceEntities)`

File: src/components/pages/devices/flyouts/deviceDelete/deviceDelete.js [hunk 6]
```diff
@@ -61,13 +67,49 @@ export class DeviceDelete extends Component {
...
+
   populateDevicesState = (devices = []) => {
-    const physicalDevices = devices.filter(({ isSimulated }) => !isSimulated);
-    this.setState({ physicalDevices, containsSimulatedDevices: (physicalDevices.length !== devices.length) });
+    const curDevices = this.state.devices || [];
+
+    if (curDevices.length !== devices.length) {
+      const physicalDevices = devices.filter(({ isSimulated }) => !isSimulated);
+      this.setState({ devices, physicalDevices, containsSimulatedDevices: (physicalDevices.length !== devices.length) });
+    }
```
Comment (spryor) [inline src/components/pages/devices/flyouts/deviceDelete/deviceDelete.js]: Merge the else and the if and put it on one line

`} else if {`

File: src/components/pages/devices/flyouts/deviceDelete/deviceDelete.js [hunk 7]
```diff
@@ -61,13 +67,49 @@ export class DeviceDelete extends Component {
...
 
-  deleteDevices = () => {
-    this.setState({ isPending: true });
+  toggleConfirm = ({ target: { value } }) => {
+    if (this.state.changesApplied) {
+      this.setState(
+        { confirmStatus: value, changesApplied: false, successCount: 0 },
+        this.populateDevicesState(this.ensureDevicesExist(this.props.devices))
+      );
+    }
```
Comment (spryor) [inline src/components/pages/devices/flyouts/deviceDelete/deviceDelete.js]: `} else {`

File: src/components/pages/devices/flyouts/deviceDelete/deviceDelete.scss [hunk 8]
```diff
@@ -8,6 +8,9 @@
 
   .device-delete-header { @include rem-fallback(padding, 12px, 0px); }
 
+  .toggle-btn-div {
```
Comment (spryor) [inline src/components/pages/devices/flyouts/deviceDelete/deviceDelete.scss]: one line

File: src/components/pages/devices/flyouts/deviceDelete/deviceDelete.scss [hunk 9]
```diff
@@ -8,6 +8,9 @@
...
+    @include rem-fallback(margin, 12px 0px);
```
Comment (spryor) [inline src/components/pages/devices/flyouts/deviceDelete/deviceDelete.scss]: add commas for the margin params

File: src/components/pages/rules/flyouts/ruleEditor/ruleEditor.scss [hunk 10]
```diff
@@ -15,4 +15,21 @@
     padding-left: 0px;
     @include rem-fallback(margin-top, 24px);
   }
+
+  .rule-editor-error {
+    border: 1px transparent solid;
+    @include rem-fallback(margin, 12px 0px);
```
Comment (spryor) [inline src/components/pages/rules/flyouts/ruleEditor/ruleEditor.scss]: missing comma after `12px`

File: src/components/pages/rules/flyouts/ruleStatus/ruleStatus.js [hunk 11]
```diff
@@ -46,12 +48,19 @@ export class RuleStatus extends Component {
     if (this.subscription) this.subscription.unsubscribe();
   }
 
-  onToggle = ({ target: { value } }) => this.setState({ status: value })
+  onToggle = ({ target: { value } }) => {
+    if (this.state.changesApplied) {
+      this.setState({ status: value, changesApplied: false });
+    }
```
Comment (spryor) [inline src/components/pages/rules/flyouts/ruleStatus/ruleStatus.js]: `} else {`

File: src/components/pages/rules/flyouts/ruleStatus/ruleStatus.scss [hunk 12]
```diff
@@ -16,4 +16,21 @@
   }
 
   .padded-bottom { @include rem-fallback(padding-bottom, 24px); }
+
+  .rule-status-error {
+    border: 1px transparent solid;
+    @include rem-fallback(margin, 12px 0px);
```
Comment (spryor) [inline src/components/pages/rules/flyouts/ruleStatus/ruleStatus.scss]: missing comma

File: src/components/pages/devices/flyouts/deviceDelete/deviceDelete.js (hunk 4 above)
Comment (mechaffin) [reply src/components/pages/devices/flyouts/deviceDelete/deviceDelete.js]: Yeah, yeah. Just a PITA to convert every time ... when debugging and having to use console.log to do it.

**Pull request merged**